py3status/modules/yubikey.py

```diff
@@ -0,0 +1,109 @@
+# -*- coding: utf-8 -*-
+"""
+Show an indicator when YubiKey is waiting for a touch.
+
+Configuration parameters:
+    format: Display format for the module.
+        (default '[YubiKey[\?if=is_gpg ][\?if=is_u2f ]]')
+    socket_path: A path to the yubikey-touch-detector socket file.
+        (default '$XDG_RUNTIME_DIR/yubikey-touch-detector.socket')
+
+Control placeholders:
+    {is_gpg} a boolean, True if YubiKey is waiting for a touch due to a gpg operation.
+    {is_u2f} a boolean, True if YubiKey is waiting for a touch due to a pam-u2f request.
+
+Requires:
+    github.com/maximbaz/yubikey-touch-detector: tool to detect when YubiKey is waiting for a touch
+
+@author Maxim Baz (https://github.com/maximbaz)
+@license BSD
+
+SAMPLE OUTPUT
+{'full_text': 'YubiKey', 'urgent': True}
+"""
+
+import time
+import os
+import socket
+import threading
+
+
+class YubiKeyTouchDetectorListener(threading.Thread):
+    """
+    A thread watchng if YubiKey is waiting for a touch
+    """
+
+    def __init__(self, parent):
+        super(YubiKeyTouchDetectorListener, self).__init__()
+        self.parent = parent
+
+    def _connect_socket(self):
+        try:
+            self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+            self.socket.connect(self.parent.socket_path)
+        except:
+            self.socket = None
+
+    def run(self):
+        while not self.parent.killed.is_set():
+            self._connect_socket()
+
+            if self.socket is None:
+                # Socket is not available, try again soon
+                time.sleep(60)
+                continue
```

Comment:
(Thought) Do we need `time.sleep(60)` here?
1) Would the thread not keep going and failing until it's finally made?
2) What happen if ssh or gpg is made, but we're in this time.sleep period? I think it could not show up... or even could show up several seconds late! 😡

Reply:
1. It's not a good idea to completely avoid `sleep`, if the socket connection cannot be established, there is likely an issue with the backend service and manual intervention is needed, and until then connection will probably fail next time as well. Trying to reconnect too often will only result if wasting lots of cycles. One minute is a sweet spot that I hardcode on purpose, not too often and yet fast enough for an automatic recovery - in real life YubiKey is used much rarely than once per minute.

However your question lead me to the following thought, since manual intervention is likely needed, perhaps silently sleeping is not enough, and we need to show an error message in red color. I saw other modules do this when they fail. Could you point me to how to achieve this?

2. It will not show up indeed, whether the problem is on the backend side (not publishing events) or on the module side (not listening to the events), no connection means no events received and no messages shown. However the notification will not show up several seconds late, when connection is finally established, it is starting from the clean state, only next even will trigger a message to be shown.

Reply:
`self.py3.error`. I think it'll keep going behind the scene until https://github.com/ultrabug/py3status/issues/1114 gets addressed. @tobes also said something about extending the `self.py3.error` exceptions to threads. 
```
./bluetooth.py:130:            self.py3.error(STRING_NOT_STARTED)
./yandexdisk_status.py:58:            self.py3.error(STRING_ERROR)
./coin_market.py:170:                self.py3.error('bad markets')
./external_script.py:76:            self.py3.error(output)
```

Reply:
I don't know this well, but quick stab...
```
diff --git a/py3status/modules/yubikey.py b/py3status/modules/yubikey.py
index 621a3c90..0d82667c 100644
--- a/py3status/modules/yubikey.py
+++ b/py3status/modules/yubikey.py
@@ -22,7 +22,6 @@ SAMPLE OUTPUT
 {'full_text': 'YubiKey', 'urgent': True}
 """
 
-import time
 import os
 import socket
 import threading
@@ -41,17 +40,16 @@ class YubiKeyTouchDetectorListener(threading.Thread):
         try:
             self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
             self.socket.connect(self.parent.socket_path)
-        except:
+        except Exception as e:
             self.socket = None
+            self.error = e
 
     def run(self):
         while not self.parent.killed.is_set():
             self._connect_socket()
 
-            if self.socket is None:
-                # Socket is not available, try again soon
-                time.sleep(60)
-                continue
+            if self.error:
+                self.parent.py3.update()
 
             while not self.parent.killed.is_set():
                 data = self.socket.recv(5)
@@ -89,6 +87,8 @@ class Py3status:
         YubiKeyTouchDetectorListener(self).start()
 
     def yubikey(self):
+        if self.error:
+            raise self.error
         response = {
             'cached_until': self.py3.CACHE_FOREVER,
             'full_text': self.py3.safe_format(self.format, self.status),
```

Reply:
awesome, thanks, it is very beautiful and descriptive now, clear to the user when this happens and what to investigate 🙂